### PR TITLE
Add HtmlEncode, UrlEncode, and JavaScriptStringEncode

### PR DESCRIPTION
**Pull request by GrabYourPitchforks**
Add HtmlEncode, UrlEncode, and JavaScriptStringEncode.
This is based loosely on the Anti-XSS encoder from WPL.

[Edited from original description.]

/cc @Eilon @Tratcher

### CODE REVIEW

**Comment by GrabYourPitchforks** [comment]
Quick stats:
- This gets around 2x the throughput of WebUtility.HtmlEncode for inputs which don't require any chars to be encoded.
- This gets around 3x the throughput of WebUtility.HtmlEncode for inputs in which at least one char must be encoded.
- The memory utilization is 8KB static data (per AD) + 8KB per HtmlEncoder instance. Compare to Anti-XSS, which utilizes approx 2-3 MB static.

It is intended that a single HtmlEncoder instance be created and reused, but applications can create as many instances as they wish in practice. (The ctor is a little slow, so I wouldn't encourage this, but it's not the end of the world if this happens.) I can also consider adding a HtmlEncoder.Instance static convenience property if developers would find it useful.

**Comment by dougbu** [comment]
like the `Instance` approach.  we create lots of `DefaultHtmlGenerator` instances that currently depend on `WebUtility.HtmlEncode()`.  would simplify the switch-over if they could just call `HtmlEncoder.Instance.HtmlEncode()`.

**Comment by GrabYourPitchforks** [comment]
@dougbu Ideally framework developers wouldn't use these APIs directly but would instead accept an instance of an IHtmlEncoder service. That allows application developers to change the default encoder system-wide.

**Comment by dougbu** [comment]
ah, hadn't noticed the service.  much more sensible.

File: src/Microsoft.AspNet.WebUtilities/Encoders/HtmlEncoder.cs [hunk 1]
```diff
@@ -0,0 +1,229 @@
...
+
+        /// <summary>
+        /// A default instance of the HtmlEncoder, equivalent to allowing only
+        /// the 'Basic Latin' character range.
+        /// </summary>
+        public static HtmlEncoder Default
+        {
+            get
+            {
+                HtmlEncoder defaultEncoder = Volatile.Read(ref _defaultEncoder);
```
Comment (davidfowl) [inline src/Microsoft.AspNet.WebUtilities/Encoders/HtmlEncoder.cs]: Any reason you didn't use LazyInitializer?

File: src/Microsoft.AspNet.WebUtilities/Encoders/HtmlEncoder.cs [hunk 2]
```diff
@@ -0,0 +1,229 @@
...
+            // The worst case encoding is 8 output chars per input char: [input] U+FFFF -> [output] "&#xFFFF;"
+            // We don't need to worry about astral code points since they consume *two* input chars to
+            // generate at most 10 output chars ("&#x10FFFF;"), which equates to 5 output per input.
+            int numCharsWhichMayRequireEncoding = input.Length - idxOfFirstCharWhichRequiresEncoding;
+            int sbCapacity = checked(idxOfFirstCharWhichRequiresEncoding + EncoderCommon.GetCapacityOfOutputStringBuilder(numCharsWhichMayRequireEncoding, worstCaseOutputCharsPerInputChar: 8));
+            Debug.Assert(sbCapacity >= input.Length);
+
+            // Allocate the StringBuilder with the first (known to not require encoding) part of the input string,
+            // then begin encoding from the last (potentially requiring encoding) part of the input string.
+            StringBuilder builder = new StringBuilder(input, 0, idxOfFirstCharWhichRequiresEncoding, sbCapacity);
```
Comment (davidfowl) [inline src/Microsoft.AspNet.WebUtilities/Encoders/HtmlEncoder.cs]: More var

File: src/Microsoft.AspNet.WebUtilities/Encoders/HtmlEncoder.cs (hunk 1 above)
Comment (GrabYourPitchforks) [inline src/Microsoft.AspNet.WebUtilities/Encoders/HtmlEncoder.cs]: Avoids static factory allocation for a code path that will be used at most once per app.

Comment (davidfowl) [inline src/Microsoft.AspNet.WebUtilities/Encoders/HtmlEncoder.cs]: Is there anywhere else we should be using this pattern? Or is it more a case by case thing? 
 /cc @Eilon

Comment (GrabYourPitchforks) [inline src/Microsoft.AspNet.WebUtilities/Encoders/HtmlEncoder.cs]: Very much a case-by-case thing that depends on how many occurrences of LazyInitializer there are and how often the accessor is likely to be called. For instance, I just stuck with a standard LazyInitializer on UnicodeFilters.None for simplicity.

**Comment by davidfowl** [comment]
We should file a bug to introduce some IHtmlEncoder interface somewhere for app consumption in the higher layers.

File: src/Microsoft.AspNet.WebUtilities/Encoders/UnicodeHelpers.cs [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+            }
+        }
+
+        private static int GetScalarValueFromUtf16SurrogatePair(char highSurrogate, char lowSurrogate)
+        {
+            Debug.Assert(Char.IsHighSurrogate(highSurrogate));
+            Debug.Assert(Char.IsLowSurrogate(lowSurrogate));
+
+            // See http://www.unicode.org/versions/Unicode6.2.0/ch03.pdf, Table 3.5 for the
+            // details of this conversion. We don't use Char.ConvertToUtf32 because its exception
```
Comment (davidfowl) [inline src/Microsoft.AspNet.WebUtilities/Encoders/UnicodeHelpers.cs]: This is interesting. We should file a bug on github.com/dotnet/coreclr

Comment (GrabYourPitchforks) [inline src/Microsoft.AspNet.WebUtilities/Encoders/UnicodeHelpers.cs]: Likely not worthwhile. Their API is perfectly fine for its intended use case; our use case is slightly different.

File: src/Microsoft.AspNet.WebUtilities/Encoders/HexUtil.cs [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+using System;
+using System.Diagnostics;
+using System.Runtime.CompilerServices;
+
+namespace Microsoft.AspNet.WebUtilities.Encoders
+{
+    /// <summary>
+    /// Contains helpers for dealing with byte-hex char conversions.
+    /// </summary>
+    internal static class HexUtil
```
Comment (dougbu) [inline src/Microsoft.AspNet.WebUtilities/Encoders/HexUtil.cs]: nit: why not `public` in an `...Internals` namespace?

File: src/Microsoft.AspNet.WebUtilities/Encoders/HexUtil.cs [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+    {
+        /// <summary>
+        /// Converts a number 0 - 15 to its associated hex character '0' - 'F'.
+        /// </summary>
+        /// <returns></returns>
```
Comment (dougbu) [inline src/Microsoft.AspNet.WebUtilities/Encoders/HexUtil.cs]: fill in `<returns/>`

File: src/Microsoft.AspNet.WebUtilities/Encoders/HtmlEncoder.cs [hunk 6]
```diff
@@ -0,0 +1,229 @@
...
+        /// </summary>
+        public HtmlEncoder()
+            : this(UnicodeFilters.BasicLatin)
+        {
+        }
+
+        /// <summary>
+        /// Instantiates an encoder using a custom allow list of characters.
+        /// </summary>
+        public HtmlEncoder(params ICodePointFilter[] filters)
```
Comment (dougbu) [inline src/Microsoft.AspNet.WebUtilities/Encoders/HtmlEncoder.cs]: nit: add `<param name="filters"/>` doc

File: src/Microsoft.AspNet.WebUtilities/Encoders/HtmlEncoder.cs [hunk 7]
```diff
@@ -0,0 +1,229 @@
...
+            {
+                if (!IsCharacterAllowed(value[i]))
+                {
+                    return HtmlEncodeImpl(value, i);
+                }
+            }
+            return value;
+        }
+
+        private string HtmlEncodeImpl(string input, int idxOfFirstCharWhichRequiresEncoding)
```
Comment (dougbu) [inline src/Microsoft.AspNet.WebUtilities/Encoders/HtmlEncoder.cs]: nit: toward avoiding abbreviations, how about `HtmlEncodeCore()`?

File: src/Microsoft.AspNet.WebUtilities/Encoders/HtmlEncoder.cs [hunk 8]
```diff
@@ -0,0 +1,229 @@
...
+        {
+            Debug.Assert(idxOfFirstCharWhichRequiresEncoding >= 0);
+            Debug.Assert(idxOfFirstCharWhichRequiresEncoding < input.Length);
+
+            // The worst case encoding is 8 output chars per input char: [input] U+FFFF -> [output] "&#xFFFF;"
+            // We don't need to worry about astral code points since they consume *two* input chars to
+            // generate at most 10 output chars ("&#x10FFFF;"), which equates to 5 output per input.
+            int numCharsWhichMayRequireEncoding = input.Length - idxOfFirstCharWhichRequiresEncoding;
+            int sbCapacity = checked(idxOfFirstCharWhichRequiresEncoding + EncoderCommon.GetCapacityOfOutputStringBuilder(numCharsWhichMayRequireEncoding, worstCaseOutputCharsPerInputChar: 8));
```
Comment (dougbu) [inline src/Microsoft.AspNet.WebUtilities/Encoders/HtmlEncoder.cs]: nit: wrap long line

File: src/Microsoft.AspNet.WebUtilities/Encoders/HtmlEncoder.cs [hunk 9]
```diff
@@ -0,0 +1,229 @@
...
+                    if (IsCharacterAllowed(c))
+                    {
+                        builder.Append(c);
+                    }
+                    else
+                    {
+                        if (c == '<') { builder.Append("&lt;"); }
+                        else if (c == '>') { builder.Append("&gt;"); }
+                        else if (c == '&') { builder.Append("&amp;"); }
+                        else if (c == '\"') { builder.Append("&quot;"); }
```
Comment (dougbu) [inline src/Microsoft.AspNet.WebUtilities/Encoders/HtmlEncoder.cs]: should this map `'\''` to `"&apos;"`?

File: src/Microsoft.AspNet.WebUtilities/Encoders/UnicodeHelpers.cs [hunk 10]
```diff
@@ -0,0 +1,164 @@
+﻿// Copyright (c) Microsoft Open Technologies, Inc. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Diagnostics;
+using System.Reflection;
+using System.Runtime.CompilerServices;
+using System.Threading;
+
+namespace Microsoft.AspNet.WebUtilities.Encoders
```
Comment (dougbu) [inline src/Microsoft.AspNet.WebUtilities/Encoders/UnicodeHelpers.cs]: nits:
- a few long lines in this class can be wrapped
- needs a bit more `var`.  same for last couple of methods in `UnicodeFilters`

File: test/Microsoft.AspNet.WebUtilities.Tests/Encoders/EncoderCommonTests.cs [hunk 11]
```diff
@@ -0,0 +1,21 @@
+﻿// Copyright (c) Microsoft Open Technologies, Inc. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using Xunit;
+
+namespace Microsoft.AspNet.WebUtilities.Encoders
+{
+    public class EncoderCommonTests
```
Comment (dougbu) [inline test/Microsoft.AspNet.WebUtilities.Tests/Encoders/EncoderCommonTests.cs]: nit: guidelines say `...Test` (singular)

File: test/Microsoft.AspNet.WebUtilities.Tests/Encoders/UnicodeFiltersTests.cs [hunk 12]
```diff
@@ -0,0 +1,984 @@
...
+            RunRangeTestImpl(UnicodeFilters.HalfwidthandFullwidthForms, first: '\uFF00', last: '\uFFEF');
+        }
+
+        [Fact]
+        public void Range_Specials()
+        {
+            RunRangeTestImpl(UnicodeFilters.Specials, first: '\uFFF0', last: '\uFFFF');
+        }
+
+        private static void RunRangeTestImpl(ICodePointFilter filter, char first, char last)
```
Comment (dougbu) [inline test/Microsoft.AspNet.WebUtilities.Tests/Encoders/UnicodeFiltersTests.cs]: nit: just `RunRangeTest()`

File: test/Microsoft.AspNet.WebUtilities.Tests/Encoders/HtmlEncoderTests.cs [hunk 13]
```diff
@@ -0,0 +1,297 @@
...
+
+namespace Microsoft.AspNet.WebUtilities.Encoders
+{
+    public class HtmlEncoderTests
+    {
+        [Fact]
+        public void Ctor_WithCustomFilters()
+        {
+            // Arrange
+            CustomCodePointFilter filter1 = new CustomCodePointFilter('a', 'b');
```
Comment (dougbu) [inline test/Microsoft.AspNet.WebUtilities.Tests/Encoders/HtmlEncoderTests.cs]: nit: test methods in general aren't `var`ish enough

File: test/Microsoft.AspNet.WebUtilities.Tests/Encoders/HtmlEncoderTests.cs [hunk 14]
```diff
@@ -0,0 +1,297 @@
...
+using System.Collections.Generic;
+using System.Globalization;
+using Xunit;
+
+namespace Microsoft.AspNet.WebUtilities.Encoders
+{
+    public class HtmlEncoderTests
+    {
+        [Fact]
+        public void Ctor_WithCustomFilters()
```
Comment (dougbu) [inline test/Microsoft.AspNet.WebUtilities.Tests/Encoders/HtmlEncoderTests.cs]: nit: avoid abbreviations even for `Constructor_...`

**Comment by dougbu** [comment]
:watch:

File: src/Microsoft.AspNet.WebUtilities/Encoders/HtmlEncoder.cs (hunk 9 above)
Comment (GrabYourPitchforks) [inline src/Microsoft.AspNet.WebUtilities/Encoders/HtmlEncoder.cs]: No - the 'apos' entity is undefined in HTML 4, so most encoders maintain compat by not emitting the entity.

File: src/Microsoft.AspNet.WebUtilities/Encoders/HexUtil.cs (hunk 5 above)
Comment (Eilon) [inline src/Microsoft.AspNet.WebUtilities/Encoders/HexUtil.cs]: This is an internal API, it doesn't need any docs at all...

**Pull request merged**